src/librustc/ty/context.rs

```diff
@@ -2848,8 +2848,29 @@ impl<'a, T, R> InternIteratorElement<T, R> for &'a T
 
 impl<T, R, E> InternIteratorElement<T, R> for Result<T, E> {
     type Output = Result<R, E>;
-    fn intern_with<I: Iterator<Item=Self>, F: FnOnce(&[T]) -> R>(iter: I, f: F) -> Self::Output {
-        Ok(f(&iter.collect::<Result<SmallVec<[_; 8]>, _>>()?))
+    fn intern_with<I: Iterator<Item=Self>, F: FnOnce(&[T]) -> R>(mut iter: I, f: F)
+            -> Self::Output {
+        // This code is hot enough that it's worth specializing for the most
+        // common length lists, to avoid the overhead of `SmallVec` creation.
+        // The match arms are in order of frequency. The 1, 2, and 0 cases are
+        // typically hit in ~95% of cases. We assume that if the upper and
```

Comment:
Is this assumption OK? It seems to run counter to the `Iterator` docs -- In particular, are you sure that no `unsafe` code in the compiler relies on the correctness of collecting? You could specialize for `TrustedLen` in which case this assumption is definitely OK.

Reply:
First I tried adding blanket `TrustedLen` requirements, but some of the iterators that feed into this method don't implement `TrustedLen`.

Then I tried to use specialization but failed. The problem is that we have this function:
```
fn intern_with<I: Iterator<Item=Self>, F: FnOnce(&[T]) -> R>(iter: I, f: F)
```
And we want one behaviour if `I` implements `TrustedLen`, and another behaviour if it doesn't. But you can't do that with an argument. This function appears within an `impl` block for `Result<T, E>`, i.e. `I` isn't part of the `impl` type.

As for the existing code... if the iterator gives fewer elements than the size hint (e.g. 1 when the hint was `(2, Some(2))`) then an `unwrap` will safely fail within the function. If the iterator gives more elements than the size hint (e.g. 2 when the hint was `(1, Some(1))`) then the wrong value will be interned. This could certainly cause correctness issues. It seems unlikely it could lead to unsafety, though I can't guarantee it. Also, it seems unlikely that an iterator would erroneously claim an exact size hint (i.e. where the lower bound and the upper bound match).

Reply:
I can't say that reasoning based on probability makes me very confident in the safety of this change. :/

How bad/expensive would it be to `assert!` that calling `next()` again yields `None`?

Reply:
I don't see any unsafe code here, and `size_hint` is required to produce conservative lower and upper bounds. If the bounds are the same, we know the size exactly.